storage/innobase/row/row0mysql.cc

```diff
@@ -2299,9 +2299,6 @@ row_insert_for_mysql_using_ins_graph(
 		return(DB_TABLE_CORRUPT);
 	}
 
-	if (UNIV_LIKELY_NULL(prebuilt->compress_heap))
-		mem_heap_empty(prebuilt->compress_heap);
-
```

Comment:
@nitendra-bhosle Can you please explain how the freed compress_heap causes update to fail (not find the record?)?

Reply:
@satya-bodapati 'mem_heap_empty' function Empties a memory heap but first memory block of the heap is not freed 
hence inside 'function row_compress_column' prebuilt->compress_heap is not NULL 
But memory is emptied.

Reply:
@nitendra-bhosle what happens after that? is the buffer insufficient to compress the column?

Reply:
and the question how this translates to update not being able to find the row still remains

Reply:
Inside method : 'row_compress_column'
buf = static_cast<byte*>(mem_heap_zalloc(
			prebuilt->compress_heap,buf_len));
this call operates on prebuilt->compress , which tries to memset memory , which it doesn't have ownership,
Which leads to memory pointed by pointer "const byte*	data" to be getting corrupted/emptied.
And hence buf returned from this method is empty .Which is causing update not being able to find the row

Reply:
how did you prove that it doesn't have ownership? 
memset() is working on new block of memory returned by heap_zalloc().

I am not sure if I understand the fix. row_mysql_prebuilt_free_compress_heap() seems to be called too.

Reply:
Inside function 'row_compress_column' 
buf = static_cast<byte*>(mem_heap_zalloc(
			prebuilt->compress_heap,buf_len));

this statement empties 'data' . Which happens when prebuilt->compress_heap doesn't have memory allocated.
I checked this by manually executing 
prebuilt->compress_heap =
			mem_heap_create(ut_max(UNIV_PAGE_SIZE, buf_len)); 
statement everythime.
----------------------------------------------
'mem_heap_create'  create heap of size maximum of UNIV_PAGE_SIZE or buf_len , which is most likely to be UNIV_PAGE_SIZE

'mem_heap_zalloc' allocates and memsets  'buf_len' bytes out of heap block created by  'mem_heap_create'

'row_mysql_prebuilt_free_compress_heap' free's heap created by 'mem_heap_create' , this executes on end of statement.

Ideally there should be pair of 'mem_heap_create' and 'row_mysql_prebuilt_free_compress_heap' for each sql statement

Reply:
row_mysql_prebuilt_free_compress_heap() is already called at end of statement. once heap is created, other iterations need not bother about creating the heap again. mem_heap_t is dynamic memory allocator. ie it will allocate new blocks requested. Something else is going @nitendra-bhosle

Reply:
Specially since this happening with partitions, check if we need separate compressed heaps for paritions. 
Check m_blob_heap_parts/ clear_blob_heaps(). Do we need something like that?

Reply:
I will explore this.

Reply:
@nitendra-bhosle 

I debugged a bit and I think your fix makes sense but here is the explanation

So the way blobs work is storage engines only return a pointer to the blob (not make a copy) to the server. This pointer is from the prebuilt->blob_heap. If blob_heap is emptied, the blob returned by SE is corrupted.

Coming to the UPDATE statement:

1. First it searches for the row and positions the cursor (the blob is decompressed and stored in blob_heap and record is returned to server)
2. Builds new row and then calls update(old_row, new_row). Since the blob field doesn't change, the blob from old row is stored in the new row (mysql record)
3. Parition hanlder detects that the updated row should be in different partition
4. Insert the new row in new parition. During insertion, we empty the blob heap. This releases the blocks used by blob and corrupts the blob data in new row
5. So we end up inserting garbage to the new partition after the update

Requests
--------
1. Update commit message explaining the sequence clearly

2. Fix this blob heap code in row_sel_store_mysql_rec()
           if (UNIV_LIKELY_NULL(prebuilt->blob_heap)) {
                row_mysql_prebuilt_free_blob_heap(prebuilt);
        }

        if (UNIV_LIKELY_NULL(prebuilt->compress_heap))
                mem_heap_empty(prebuilt->compress_heap);
  
         call free for compress_heap as well
        
3. Bring the replication testcase from PS-7599

Reply:
4. Run the mtr testcase on valgrind/ASAN and paste the result here.